src/vos/sys_db.c

```diff
@@ -0,0 +1,435 @@
+/** private information of VOS system DB (pool & container) */
+struct vos_sys_db {
+	/** exported part of VOS system DB */
+	struct sys_db		 db_pub;
+	char			*db_file;
+	char			*db_path;
+	struct umem_instance	*db_umm;
+	/* DB should be destroyed on exit */
+	bool			 db_vos_self;
+	ABT_mutex		 db_lock;
+	uuid_t			 db_pool;
+	uuid_t			 db_cont;
+	daos_handle_t		 db_poh;
+	daos_handle_t		 db_coh;
+	daos_unit_oid_t		 db_obj;
+};
+
+static struct vos_sys_db	vos_db;
+
+/** data structure for VOS I/O */
+struct sys_db_io {
+	d_iov_t			io_key;
+	daos_iod_t		io_iod;
+	d_sg_list_t		io_sgl;
+};
+
+static int
+db_upsert(struct sys_db *db, char *table, d_iov_t *key, d_iov_t *val);
+
+static int
+db_fetch(struct sys_db *db, char *table, d_iov_t *key, d_iov_t *val);
+
+struct vos_sys_db *
+db2vos(struct sys_db *db)
+{
+	return container_of(db, struct vos_sys_db, db_pub);
+}
+
+static void
+db_close(struct sys_db *db)
+{
+	struct vos_sys_db *vdb = db2vos(db);
+
+	if (!daos_handle_is_inval(vdb->db_coh)) {
+		vos_cont_close(vdb->db_coh);
+		vdb->db_coh = DAOS_HDL_INVAL;
+	}
+
+	if (!daos_handle_is_inval(vdb->db_poh)) {
+		vos_pool_close(vdb->db_poh);
+		vdb->db_poh = DAOS_HDL_INVAL;
+	}
+}
+
+static void
+db_unlink(struct sys_db *db)
+{
+	struct vos_sys_db *vdb = db2vos(db);
+
+	unlink(vdb->db_file); /* ignore error code */
+}
+
+/* open or create system DB stored in pmemfile */
+static int
+db_open_create(struct sys_db *db, bool try_create)
+{
+	struct vos_sys_db *vdb = db2vos(db);
+	d_iov_t		   key;
+	d_iov_t		   val;
+	uint32_t	   ver;
+	int		   rc;
+
+	if (try_create) {
+		rc = mkdir(vdb->db_path, 0777);
+		if (rc < 0 && errno != EEXIST) {
+			rc = daos_errno2der(errno);
+			goto failed;
+		}
+
+		rc = vos_pool_create(vdb->db_file, vdb->db_pool,
+				     SYS_DB_SIZE, 0);
+		if (rc) {
+			D_CRIT("sys pool create error: "DF_RC"\n", DP_RC(rc));
+			goto failed;
+		}
+	}
+	rc = vos_pool_open(vdb->db_file, vdb->db_pool, false, &vdb->db_poh);
+	if (rc) {
+		if (rc != -DER_NONEXIST || try_create)
+			D_CRIT("sys pool open error: "DF_RC"\n", DP_RC(rc));
+		goto failed;
+	}
+
+	if (try_create) {
+		rc = vos_cont_create(vdb->db_poh, vdb->db_cont);
+		if (rc) {
+			D_CRIT("sys cont create error: "DF_RC"\n", DP_RC(rc));
+			goto failed;
+		}
+	}
+	rc = vos_cont_open(vdb->db_poh, vdb->db_cont, &vdb->db_coh);
+	if (rc) {
+		D_CRIT("sys cont open error: "DF_RC"\n", DP_RC(rc));
+		goto failed;
+	}
+
+	vdb->db_umm = vos_pool2umm(vos_hdl2pool(vdb->db_poh));
+	d_iov_set(&key, SYS_DB_MD_VER, strlen(SYS_DB_MD_VER));
+	d_iov_set(&val, &ver, sizeof(ver));
+	if (try_create) {
+		ver = SYS_DB_VERSION;
+		rc = db_upsert(db, SYS_DB_MD, &key, &val);
+		if (rc) {
+			D_CRIT("Failed to set version for sysdb: "DF_RC"\n",
+			       DP_RC(rc));
+			goto failed;
+		}
+	} else {
+		rc = db_fetch(db, SYS_DB_MD, &key, &val);
+		if (rc) {
+			D_CRIT("Failed to read sysdb version: "DF_RC"\n",
+			       DP_RC(rc));
+			goto failed;
+		}
+
+		if (ver <SYS_DB_VERSION_1 || ver > SYS_DB_VERSION) {
```

Comment:
```suggestion
		if (ver < SYS_DB_VERSION_1 || ver > SYS_DB_VERSION) {
```

Reply:
Can wait on this to another patch since this one needs to land...but if you need to update it, fix it